Additional documentation for CWL integration (#2798)
DataBiosphere/toil: docs/running/cwl.rst

Running CWL Locally
-------------------

The `toil-cwl-runner` command provides cwl-parsing functionality using cwltool, and leverages the job-scheduling and
batch system support of Toil.

To run in local batch mode, provide the CWL file and the input object file::

    $ toil-cwl-runner example.cwl example-job.yml

For a simple example of CWL with Toil see :ref:`cwlquickstart`.

Detailed Usage Instructions
---------------------------

Help information can be found by using this toil command:
::

    $ toil-cwl-runner -h

A more detailed example shows how we can specify both Toil and cwltool arguments for our workflow:
::

    $ toil-cwl-runner \
        --singularity \
        --jobStore my_jobStore \
        --batchSystem lsf \
        --workDir `pwd` \
        --outdir `pwd` \
        --logFile cwltoil.log \
        --writeLogs `pwd` \
        --logLevel DEBUG \
        --retryCount 2 \
        --disableCaching \
        --maxLogFileSize 20000000000 \
        --stats \
        standard_bam_processing.cwl \
        inputs.yaml

In this example, we set the following options, which are all passed to Toil:
::

>     --singularity: Specifies that all jobs should be run with Singularity containers instead of

Reviewer: ```suggestion
    --singularity: Specifies that all jobs should be run using the Singularity container engine instead of
```

    Docker.

    --jobStore: Path to a folder that already exists, which will contain the Toil jobstore and all
        related job-tracking information.

    --batchSystem: Use the specified HPC or Cloud-based cluster platform.

>     --workDir: The directory where all temporary files will be created for the workflow. This can be referenced as

Reviewer: ```suggestion
    --workDir: The directory where all temporary files will be created for the workflow. A subdirectory of this will be set
    as the $TMPDIR environment variable and is also available using the CWL expression 
```

        $(runtime.tmpdir) in CWL tools and workflows.

>     --outdir: Directory where final outputs will be written.

Reviewer: ```suggestion
    --outdir: Directory where final File and Directory outputs will be written. Other output types will be represented in the
                 JSON object written to stdout.
```

    --logFile: Path to the main logfile with logs from all jobs.

>     --writeLogs: Directory where all job logs will be stored to the current working directory.

Reviewer: ```suggestion
    --writeLogs: Directory where all job logs will be stored.
```

    --retryCount: How many times to retry each Toil job.

>     --disableCaching: Required for LSF batch system.

Reviewer: ```suggestion
    --disableCaching: Currently required for batch systems (LSF, slurm, gridengine, htcondor, torque)
```

    --maxLogFileSize: Logs that get larger than this value will be truncated.

>     --stats: Save resources usages in json files that can be collected with the toil stats command after the workflow is done.

Reviewer: ```suggestion
    --stats: Save resources usages in JSON files that can be collected with the toil stats command after the workflow is done.
```

samples inputs, it could look something like:

.. literalinclude:: ../../src/toil/test/docs/scripts/tutorial_cwlexample.py

Toil & CWL Tips
---------------

See logs for just one job (by using the full log file, requires knowing the job's toil-generated ID):
::

    cat cwltoil.log | grep jobVM1fIs

Grep for full commands from toil logs

This gives you a more concise view of the commands being run (note that this information is only available from
Toil when running with `--logDebug`, or when using the `pipeline_submit` script with `--logLevel DEBUG`):
::

    pcregrep -M "\[job .*\.cwl.*$\n(.*        .*$\n)*" cwltoil.log
              ^allows for multiline matching

Find Bams that have been generated for specific step while pipeline is running:
::

    find . | grep -P '^./out_tmpdir.*_MD\.bam$'

Print graphical representation of the workflow:
::

>     cwltool --print-dot workflows/innovation_pipeline.cwl | dot -Tpdf -o innovation_pipeline.pdf && open innovation_pipeline.pdf

Reviewer: I think `open` is macOS specific? `xdg-open` is more generic

Author: 👍

samples inputs, it could look something like:

.. literalinclude:: ../../src/toil/test/docs/scripts/tutorial_cwlexample.py

Toil & CWL Tips
---------------

**See logs for just one job by using the full log file**

This requires knowing the job's toil-generated ID, which can be found in the log files.
::

    cat cwltoil.log | grep jobVM1fIs

**Grep for full tool commands from toil logs**

This gives you a more concise view of the commands being run (note that this information is only available from
Toil when running with `--logDebug`).
::

    pcregrep -M "\[job .*\.cwl.*$\n(.*        .*$\n)*" cwltoil.log
              ^allows for multiline matching

**Find Bams that have been generated for specific step while pipeline is running:**
::

    find . | grep -P '^./out_tmpdir.*_MD\.bam$'

**See what jobs have been run**
::

    cat log/cwltoil.log | grep -oP "\[job .*.cwl\]" | sort | uniq

or:
::

    cat log/cwltoil.log | grep -i "issued job"

**Get status of a workflow**
::

>     $ toil status /home/johnsoni/TEST_RUNS_3/TEST_run/tmp/jobstore-09ae0acc-c800-11e8-9d09-70106fb1697e

Reviewer: ```suggestion
    $ toil status /home/user/jobstore-09ae0acc-c800-11e8-9d09-70106fb1697e
```

Match the paths

The new log file will be named something like:
::

>     file:<path to cwl tool>.cwl_<job ID>.log

Reviewer: Can you keep both versions? It is nice to see a real filename here

Author: 👍
